include/ceed/jit-source/cuda/cuda-ref-basis-tensor-at-points.h

```diff
@@ -0,0 +1,319 @@
+
+          // Contract along middle index
+          for (CeedInt k = i; k < writeLen; k += blockDim.x) {
+            const CeedInt c  = k % post;
+            const CeedInt j  = (k / post) % Q;
+            const CeedInt a  = k / (post * Q);
+            CeedScalar    vk = 0;
+
+            for (CeedInt b = 0; b < Q; b++) vk += chebyshev_x[b] * in[(a * Q + b) * post + c];
+            out[k] = vk;
+          }
+          post *= 1;
+        }
+      }
+    }
+  }
+}
+
+extern "C" __global__ void InterpTransposeAtPoints(const CeedInt num_elem, const CeedScalar *__restrict__ chebyshev_interp_1d,
+                                                   const CeedScalar *__restrict__ coords, const CeedScalar *__restrict__ u,
+                                                   CeedScalar *__restrict__ v) {
+  const CeedInt i = threadIdx.x;
+
+  __shared__ CeedScalar s_mem[BASIS_Q_1D * BASIS_P_1D + 3 * BASIS_BUF_LEN];
+  CeedScalar           *s_chebyshev_interp_1d = s_mem;
+  CeedScalar           *s_buffer_1            = s_mem + BASIS_Q_1D * BASIS_P_1D;
+  CeedScalar           *s_buffer_2            = s_buffer_1 + BASIS_BUF_LEN;
+  CeedScalar           *s_chebyshev_coeffs    = s_buffer_2 + BASIS_BUF_LEN;
+  CeedScalar            s_chebyshev_x[BASIS_DIM * BASIS_BUF_LEN];
+  for (CeedInt k = i; k < BASIS_Q_1D * BASIS_P_1D; k += blockDim.x) {
+    s_chebyshev_interp_1d[k] = interp_1d[k];
+  }
+
+  const CeedInt P             = BASIS_P_1D;
+  const CeedInt Q             = BASIS_Q_1D;
+  const CeedInt stride_0      = 1;
+  const CeedInt stride_1      = BASIS_P_1D;
+  const CeedInt u_stride      = BASIS_NUM_PTS;
+  const CeedInt v_stride      = BASIS_NUM_NODES;
+  const CeedInt u_comp_stride = num_elem * BASIS_NUM_PTS;
+  const CeedInt v_comp_stride = num_elem * BASIS_NUM_NODES;
+  const CeedInt u_size        = BASIS_NUM_PTS;
+
+  // Apply basis element by element
+  for (CeedInt elem = blockIdx.x; elem < num_elem; elem += gridDim.x) {
+    for (CeedInt comp = 0; comp < BASIS_NUM_COMP; comp++) {
+      const CeedScalar *cur_u = u + elem * u_stride + comp * u_comp_stride;
+      CeedScalar       *cur_v = v + elem * v_stride + comp * v_comp_stride;
+      CeedInt           pre   = 1;
+      CeedInt           post  = 1;
+
+      // Map from point
+      for (CeedInt k = i; k < u_size; k += blockDim.x) {
+        s_buffer_1[k] = cur_u[k];
+      }
+      for (CeedInt p = 0; p < BASIS_NUM_PTS; p++) {
+        for (CeedInt d = 0; d < BASIS_DIM; d++) {
+          __syncthreads();
+          // Update buffers used
+          pre /= 1;
+          const CeedScalar *in       = d % 2 ? s_buffer_2 : s_buffer_1;
+          CeedScalar       *out      = d == BASIS_DIM - 1 ? s_chebyshev_coeffs : (d % 2 ? s_buffer_1 : s_buffer_2);
+          const CeedInt     writeLen = pre * post * Q;
+
+          // Build Chebyshev polynomial values
+          ChebyshevPolynomialsAtPoint<BASIS_Q_1D>(coords + elem * u_stride + d * u_comp_stride, chebyshev_x);
```

Comment:
I don't think this should be shared memory as its local to the point being evaluated?

Reply:
Yeah, anything local to the point would have no reason to be shared.

Reply:
Awesome, good to know I'm not fully crazy here